Exception 'Load access fault'

## Bug description

Every few days, while booting my program, this happens just after turning on BLE:
```
5.848220 [INFO ] esp-radio configuration Config { rx_queue_size: 5, tx_queue_size: 3, static_rx_buf_num: 10, dynamic_rx_buf_num: 32, static_tx_buf_num: 0, dynamic_tx_buf_num: 32, ampdu_rx_enable: true, ampdu_tx_enable: true, amsdu_tx_enable: false, rx_ba_win: 6, max_burst_size: 1, country_code: "CN", country_code_operating_class: 0, mtu: 1492, listen_interval: 3, beacon_timeout: 6, ap_beacon_timeout: 300, failure_retry_cnt: 1, scan_method: 0 } (esp_radio esp-radio/src/lib.rs:278)
5.903834 [ERROR] panicked at 'Exception 'Load access fault' mepc=420ef8f0, mtval=ffd0000f
TrapFrame { ra: 1108277650, t0: 202, t1: 1108942680, t2: 32, t3: 0, t4: 0, t5: 573046784, t6: 14680064, a0: 8, a1: 1082462208, a2: 2, a3: 1082195356, a4: 1, a5: 1611325440, a6: 1082215564, a7: 0 }' (esp_hal src/exception_handler/mod.rs:46)
5.904106 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904129 [ERROR] ====================== PANIC ====================== (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904151 [ERROR] panicked at src/main.rs:43:5 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904244 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904265 [ERROR] Backtrace: (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904287 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904343 [ERROR] 0x420fdc4e (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904378 [ERROR] 0x42112d46 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904413 [ERROR] 0x40800494 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904448 [ERROR] 0x40800348 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904483 [ERROR] 0x420ef940 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904518 [ERROR] 0x420ef862 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904553 [ERROR] 0x420fdc22 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904588 [ERROR] 0x420ef688 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904626 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904647 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.904669 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
Firmware exited with: Application exit (Aborted)
Core 0
    Frame 0: syscall_readonly @ 0x42188f64 inline
       .cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/sys/arm_compat/syscall/riscv.rs:35:9
    Frame 1: sys_exit_extended @ 0x0000000042188f40 inline
       .cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/sys/arm_compat/mod.rs:191:9
    Frame 2: exit @ 0x0000000042188f40 inline
       .cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/sys/arm_compat/mod.rs:160:5
    Frame 3: exit @ 0x0000000042188f40
       .cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/process.rs:57:5
    Frame 4: abort @ 0x42188f9c
       .cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/process.rs:69:5
    Frame 5: abort @ 0x4210ec4e
       .cargo/git/checkouts/esp-hal-f15fbdeaa22f7411/efb4ea4/esp-backtrace/src/lib.rs:249:13
    Frame 6: panic_handler @ 0x4210ebfe
       .cargo/git/checkouts/esp-hal-f15fbdeaa22f7411/efb4ea4/esp-backtrace/src/lib.rs:120:5
    Frame 7: panic_fmt @ 0x421093ce
       .rustup/toolchains/nightly-2025-05-13-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:75:14
    Frame 8: panic @ 0x420fdc50
       init-esp32-c6/src/main.rs:43:5
    Frame 9: panic @ 0x42112d48 inline
       .cargo/registry/src/index.crates.io-1949cf8c6b5b557f/defmt-1.0.1/src/export/mod.rs:133:14
    Frame 10: ExceptionHandler @ 0x0000000042112d42
       .cargo/git/checkouts/esp-hal-f15fbdeaa22f7411/efb4ea4/esp-hal/src/exception_handler/mod.rs:0
    Frame 11: panic @ 0x42112d48 inline
       .cargo/registry/src/index.crates.io-1949cf8c6b5b557f/defmt-1.0.1/src/export/mod.rs:133:14
    Frame 12: ExceptionHandler @ 0x0000000042112d42
       .cargo/git/checkouts/esp-hal-f15fbdeaa22f7411/efb4ea4/esp-hal/src/exception_handler/mod.rs:0
```

## To Reproduce

Good question, happens every few days, once, then it's gone for a while. Haven't found a way to replicate it or appear it more often

## Environment

- Target device: esp32c6
- Crate name and version: esp-hal at commit `085beda621fb39a97b81fcb43b11b9b832d89278` (So like 3 days old)

Well it's not reliable to reproduce in my program (Maybe would be if probe-rs attach wouldn't crash the device sometimes...), so I doubt trying to create a reproductible example is possible

Anyway, another weird BLE error, this is a new one, I never encrountered it:
```
5.850725 [INFO ] esp-radio configuration Config { rx_queue_size: 5, tx_queue_size: 3, static_rx_buf_num: 10, dynamic_rx_buf_num: 32, static_tx_buf_num: 0, dynamic_tx_buf_num: 32, ampdu_rx_enable: true, ampdu_tx_enable: true, amsdu_tx_enable: false, rx_ba_win: 6, max_burst_size: 1, country_code: "CN", country_code_operating_class: 0, mtu: 1492, listen_interval: 3, beacon_timeout: 6, ap_beacon_timeout: 300, failure_retry_cnt: 1, scan_method: 0 } (esp_radio esp-radio/src/lib.rs:279)
5.915551 [ERROR] panicked at 'ASSERT r_ble_hci_trans_hs_cmd_tx:214 0 0' (esp_radio src/ble/npl.rs:423)
5.915730 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.915752 [ERROR] ====================== PANIC ====================== (esp_backtrace esp-backtrace/src/lib.rs:70)
5.915774 [ERROR] panicked at src/main.rs:44:5 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.915870 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.915891 [ERROR] Backtrace: (esp_backtrace esp-backtrace/src/lib.rs:70)
5.915913 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.915966 [ERROR] 0x420ffd16 (esp_backtrace esp-backtrace/src/lib.rs:70)
5.916007 [ERROR] 0x4211a01c (esp_backtrace esp-backtrace/src/lib.rs:70)
5.916045 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.916066 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
5.916088 [ERROR]  (esp_backtrace esp-backtrace/src/lib.rs:70)
Firmware exited with: Application exit (Aborted)
ERROR probe_rs_debug::debug_info: Other("UNWIND: Tried to unwind `RegisterRule` at CFA = None.")
Core 0
    Frame 0: syscall_readonly @ 0x4218c354 inline
       ~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/sys/arm_compat/syscall/riscv.rs:35:9
    Frame 1: sys_exit_extended @ 0x000000004218c330 inline
       ~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/sys/arm_compat/mod.rs:191:9
    Frame 2: exit @ 0x000000004218c330 inline
       ~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/sys/arm_compat/mod.rs:160:5
    Frame 3: exit @ 0x000000004218c330
       ~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/process.rs:57:5
    Frame 4: abort @ 0x4218c38c
       ~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.20/src/process.rs:69:5
    Frame 5: abort @ 0x42110cea
       ~/.cargo/git/checkouts/esp-hal-f15fbdeaa22f7411/677a933/esp-backtrace/src/lib.rs:249:13
    Frame 6: panic_handler @ 0x42110c9a
       ~/.cargo/git/checkouts/esp-hal-f15fbdeaa22f7411/677a933/esp-backtrace/src/lib.rs:120:5
    Frame 7: panic_fmt @ 0x4210b472
       ~/.rustup/toolchains/nightly-2025-05-13-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:75:14
    Frame 8: panic @ 0x420ffd18
       coolpathinit/init-esp32-c6/src/main.rs:44:5
    Frame 9: panic @ 0x4211a01e inline
       ~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/defmt-1.0.1/src/export/mod.rs:133:14
    Frame 10: osi_assert @ 0x000000004211a018
       ~/.cargo/git/checkouts/esp-hal-f15fbdeaa22f7411/677a933/esp-radio/src/ble/npl.rs:423:9
    Frame 11: <unknown function @ 0x4219aa6a> @ 0x4219aa6a

    Frame 12: <unknown function @ 0x4219aa6a> @ 0x4219aa6a

    Frame 13: <unknown function @ 0x4219aa6a> @ 0x4219aa6a

    Frame 14: <unknown function @ 0x0000fffe> @ 0x0000fffe

    Frame 15: <unknown function @ 0x4085401e> @ 0x4085401e

    Frame 16: {async_fn#0} @ 0x42158d96 inline
       coolpathlogic/src/lib.rs
    Frame 17: poll<logic::__led_task_task::{async_fn_env#0}> : ERROR: UNWIND: Tried to unwind `RegisterRule` at CFA = None. @ 0x0000000042158ce0
       ~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/embassy-executor-0.7.0/src/raw/mod.rs:214:15
```

--- a/esp-radio/Cargo.toml
+++ b/esp-radio/Cargo.toml
@@ -90,15 +90,15 @@ esp-wifi-sys-esp32h2 = { version = "0.2.0", optional = true }
 esp-wifi-sys-esp32s2 = { version = "0.2.0", optional = true }
 esp-wifi-sys-esp32s3 = { version = "0.2.0", optional = true }
 
-esp32   = { version = "0.40", features = ["critical-section"], optional = true }
+esp32   = { version = "0.40", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
 esp32c2 = { version = "0.29", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
-esp32c3 = { version = "0.32", features = ["critical-section"], optional = true }
-esp32c5 = { version = "0.2",  features = ["critical-section"], optional = true }
-esp32c6 = { version = "0.23", features = ["critical-section"], optional = true }
-esp32c61 = { version = "0.3", features = ["critical-section"], optional = true }
-esp32h2 = { version = "0.19", features = ["critical-section"], optional = true }
-esp32s2 = { version = "0.31", features = ["critical-section"], optional = true }
-esp32s3 = { version = "0.35", features = ["critical-section"], optional = true }
+esp32c3 = { version = "0.32", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
+esp32c5 = { version = "0.2",  features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
+esp32c6 = { version = "0.23", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
+esp32c61 = { version = "0.3", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
+esp32h2 = { version = "0.19", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
+esp32s2 = { version = "0.31", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
+esp32s3 = { version = "0.35", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
 
 # Optional dependencies enabling ecosystem features
 embedded-io-06 = { package = "embedded-io", version = "0.6", default-features = false, optional = true }

--- a/esp-radio/Cargo.toml
+++ b/esp-radio/Cargo.toml
@@ -91,7 +91,7 @@ esp-wifi-sys-esp32s2 = { version = "0.2.0", optional = true }
 esp-wifi-sys-esp32s3 = { version = "0.2.0", optional = true }
 
 esp32   = { version = "0.40", features = ["critical-section"], optional = true }
-esp32c2 = { version = "0.29", features = ["critical-section"], optional = true }
+esp32c2 = { version = "0.29", features = ["critical-section"], optional = true, git = "https://github.com/esp-rs/esp-pacs", rev = "9504df468a04b67b960dc5cc7430788efa257207" }
 esp32c3 = { version = "0.32", features = ["critical-section"], optional = true }
 esp32c5 = { version = "0.2",  features = ["critical-section"], optional = true }
 esp32c6 = { version = "0.23", features = ["critical-section"], optional = true }

--- a/esp-radio/src/common_adapter.rs
+++ b/esp-radio/src/common_adapter.rs
@@ -422,18 +422,18 @@ pub(crate) fn enable_wifi_power_domain() {
     // subsystems in their previous state.
     #[cfg(esp32c2)]
     {
-        const WIFIBB_RST: u32 = 1 << 0;
-        const FE_RST: u32 = 1 << 1;
-        const WIFIMAC_RST: u32 = 1 << 2;
-        const BLE_RPA_RST: u32 = 1 << 27;
-        const MODEM_RESET: u32 = WIFIBB_RST | FE_RST | WIFIMAC_RST | BLE_RPA_RST;
-
-        regs!(APB_CTRL)
-            .wifi_rst_en()
-            .modify(|r, w| unsafe { w.bits(r.bits() | MODEM_RESET) });
-        regs!(APB_CTRL)
-            .wifi_rst_en()
-            .modify(|r, w| unsafe { w.bits(r.bits() & !MODEM_RESET) });
+        regs!(APB_CTRL).wifi_rst_en().modify(|_, w| {
+            w.wifibb_rst().set_bit();
+            w.fe_rst().set_bit();
+            w.mac_rst().set_bit();
+            w.ble_rpa_rst().set_bit()
+        });
+        regs!(APB_CTRL).wifi_rst_en().modify(|_, w| {
+            w.wifibb_rst().clear_bit();
+            w.fe_rst().clear_bit();
+            w.mac_rst().clear_bit();
+            w.ble_rpa_rst().clear_bit()
+        });
     }
 }
 

--- a/esp-radio/src/common_adapter.rs
+++ b/esp-radio/src/common_adapter.rs
@@ -373,67 +373,59 @@ pub(crate) fn enable_wifi_power_domain() {
     // baseband and MAC). On PMU-based chips a system reset restarts the digital core
     // but leaves the modem power domain untouched, so without this the radio
     // would inherit whatever state the previously-used radio left.
-    #[cfg(any(esp32c5, esp32c6, esp32c61))]
-    {
-        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
-            w.rst_fe().set_bit();
-            w.rst_btbb().set_bit();
-            w.rst_btbb_apb().set_bit();
-            w.rst_btmac().set_bit();
-            w.rst_btmac_apb().set_bit();
-            w.rst_wifibb().set_bit();
-            w.rst_wifimac().set_bit()
-        });
-        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
-            w.rst_fe().clear_bit();
-            w.rst_btbb().clear_bit();
-            w.rst_btbb_apb().clear_bit();
-            w.rst_btmac().clear_bit();
-            w.rst_btmac_apb().clear_bit();
-            w.rst_wifibb().clear_bit();
-            w.rst_wifimac().clear_bit()
-        });
-    }
-
-    // ESP32-H2 has no Wi-Fi baseband/MAC; reset the BT + RF frontend (and the
-    // 802.15.4 MAC) modem subsystems instead.
-    #[cfg(esp32h2)]
-    {
-        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
-            w.rst_fe().set_bit();
-            w.rst_btbb().set_bit();
-            w.rst_btbb_apb().set_bit();
-            w.rst_btmac().set_bit();
-            w.rst_btmac_apb().set_bit();
-            w.rst_zbmac().set_bit()
-        });
-        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
-            w.rst_fe().clear_bit();
-            w.rst_btbb().clear_bit();
-            w.rst_btbb_apb().clear_bit();
-            w.rst_btmac().clear_bit();
-            w.rst_btmac_apb().clear_bit();
-            w.rst_zbmac().clear_bit()
-        });
-    }
-
-    // ESP32-C2 has no separate modem power domain (RTC_CNTL lacks
-    // wifi_force_pd/iso), but a system reset still leaves the shared modem
-    // subsystems in their previous state.
-    #[cfg(esp32c2)]
-    {
-        regs!(APB_CTRL).wifi_rst_en().modify(|_, w| {
-            w.wifibb_rst().set_bit();
-            w.fe_rst().set_bit();
-            w.mac_rst().set_bit();
-            w.ble_rpa_rst().set_bit()
-        });
-        regs!(APB_CTRL).wifi_rst_en().modify(|_, w| {
-            w.wifibb_rst().clear_bit();
-            w.fe_rst().clear_bit();
-            w.mac_rst().clear_bit();
-            w.ble_rpa_rst().clear_bit()
-        });
+    cfg_select! {
+        any(esp32c5, esp32c6, esp32c61, esp32h2) => {
+            regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
+                w.rst_fe().set_bit();
+                w.rst_btbb().set_bit();
+                w.rst_btbb_apb().set_bit();
+                w.rst_btmac().set_bit();
+                w.rst_btmac_apb().set_bit();
+                cfg_select! {
+                    any(esp32c5, esp32c6, esp32c61) => {
+                        w.rst_wifibb().set_bit();
+                        w.rst_wifimac().set_bit()
+                    }
+                    esp32h2 => {
+                        w.rst_zbmac().set_bit()
+                    }
+                }
+            });
+            regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
+                w.rst_fe().clear_bit();
+                w.rst_btbb().clear_bit();
+                w.rst_btbb_apb().clear_bit();
+                w.rst_btmac().clear_bit();
+                w.rst_btmac_apb().clear_bit();
+                cfg_select! {
+                    any(esp32c5, esp32c6, esp32c61) => {
+                        w.rst_wifibb().clear_bit();
+                        w.rst_wifimac().clear_bit()
+                    }
+                    esp32h2 => {
+                        w.rst_zbmac().clear_bit()
+                    }
+                }
+            });
+        }
+        // ESP32-C2 has no separate modem power domain (RTC_CNTL lacks
+        // wifi_force_pd/iso), but a system reset still leaves the shared modem
+        // subsystems in their previous state.
+        esp32c2 => {
+            regs!(APB_CTRL).wifi_rst_en().modify(|_, w| {
+                w.wifibb_rst().set_bit();
+                w.fe_rst().set_bit();
+                w.mac_rst().set_bit();
+                w.ble_rpa_rst().set_bit()
+            });
+            regs!(APB_CTRL).wifi_rst_en().modify(|_, w| {
+                w.wifibb_rst().clear_bit();
+                w.fe_rst().clear_bit();
+                w.mac_rst().clear_bit();
+                w.ble_rpa_rst().clear_bit()
+            });
+        }
+        _ => {}
     }
 }
 

--- a/esp-radio/src/common_adapter.rs
+++ b/esp-radio/src/common_adapter.rs
@@ -368,6 +368,73 @@ pub(crate) fn enable_wifi_power_domain() {
             .dig_iso()
             .modify(|_, w| w.wifi_force_iso().clear_bit());
     }
+
+    // Pulse a reset over the shared modem subsystems (RF frontend, BT/Wi-Fi
+    // baseband and MAC). On PMU-based chips a system reset restarts the digital core
+    // but leaves the modem power domain untouched, so without this the radio
+    // would inherit whatever state the previously-used radio left.
+    #[cfg(any(esp32c5, esp32c6, esp32c61))]
+    {
+        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
+            w.rst_fe().set_bit();
+            w.rst_btbb().set_bit();
+            w.rst_btbb_apb().set_bit();
+            w.rst_btmac().set_bit();
+            w.rst_btmac_apb().set_bit();
+            w.rst_wifibb().set_bit();
+            w.rst_wifimac().set_bit()
+        });
+        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
+            w.rst_fe().clear_bit();
+            w.rst_btbb().clear_bit();
+            w.rst_btbb_apb().clear_bit();
+            w.rst_btmac().clear_bit();
+            w.rst_btmac_apb().clear_bit();
+            w.rst_wifibb().clear_bit();
+            w.rst_wifimac().clear_bit()
+        });
+    }
+
+    // ESP32-H2 has no Wi-Fi baseband/MAC; reset the BT + RF frontend (and the
+    // 802.15.4 MAC) modem subsystems instead.
+    #[cfg(esp32h2)]
+    {
+        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
+            w.rst_fe().set_bit();
+            w.rst_btbb().set_bit();
+            w.rst_btbb_apb().set_bit();
+            w.rst_btmac().set_bit();
+            w.rst_btmac_apb().set_bit();
+            w.rst_zbmac().set_bit()
+        });
+        regs!(MODEM_SYSCON).modem_rst_conf().modify(|_, w| {
+            w.rst_fe().clear_bit();
+            w.rst_btbb().clear_bit();
+            w.rst_btbb_apb().clear_bit();
+            w.rst_btmac().clear_bit();
+            w.rst_btmac_apb().clear_bit();
+            w.rst_zbmac().clear_bit()
+        });
+    }
+
+    // ESP32-C2 has no separate modem power domain (RTC_CNTL lacks
+    // wifi_force_pd/iso), but a system reset still leaves the shared modem
+    // subsystems in their previous state.
+    #[cfg(esp32c2)]
+    {
+        const WIFIBB_RST: u32 = 1 << 0;
+        const FE_RST: u32 = 1 << 1;
+        const WIFIMAC_RST: u32 = 1 << 2;
+        const BLE_RPA_RST: u32 = 1 << 27;
+        const MODEM_RESET: u32 = WIFIBB_RST | FE_RST | WIFIMAC_RST | BLE_RPA_RST;
+
+        regs!(APB_CTRL)
+            .wifi_rst_en()
+            .modify(|r, w| unsafe { w.bits(r.bits() | MODEM_RESET) });
+        regs!(APB_CTRL)
+            .wifi_rst_en()
+            .modify(|r, w| unsafe { w.bits(r.bits() & !MODEM_RESET) });
+    }
 }
 
 /// **************************************************************************

--- a/qa-test/Cargo.toml
+++ b/qa-test/Cargo.toml
@@ -10,7 +10,7 @@ embassy-executor = "0.10.0"
 embassy-time = "0.5.0"
 embassy-futures = "0.1"
 embassy-net = { version = "0.9.0", features = [ "tcp", "udp", "dhcpv4", "medium-ethernet"] }
-embassy-sync = "0.8.0"
+embassy-sync = "0.7.0" # TODO: update to 0.8 once trouble supports it
 embedded-graphics = "0.8.1"
 embedded-hal-async = "1.0.0"
 embedded-io-async = "0.7.0"

--- a/qa-test/src/bin/wifi_ble_modem_handoff.rs
+++ b/qa-test/src/bin/wifi_ble_modem_handoff.rs
@@ -1,0 +1,297 @@
+//! Alternating Wi-Fi / BLE radio stress test
+//!
+//! https://github.com/esp-rs/esp-hal/pull/5670
+//% FEATURES: esp-radio esp-radio/wifi esp-radio/ble esp-radio/unstable esp-hal/unstable
+//% CHIPS: esp32 esp32c3 esp32c5 esp32c6 esp32s3
+
+#![no_std]
+#![no_main]
+#![allow(static_mut_refs)]
+
+use embassy_executor::Spawner;
+use embassy_futures::{join::join, select::select};
+use embassy_time::{Duration, Timer};
+use esp_alloc as _;
+use esp_backtrace as _;
+use esp_hal::{
+    clock::CpuClock,
+    interrupt::software::SoftwareInterruptControl,
+    ram,
+    rng::Rng,
+    system::software_reset,
+    timer::timg::TimerGroup,
+};
+use esp_radio::{
+    ble::controller::BleConnector,
+    wifi::{Config, ControllerConfig, WifiController, ap::AccessPointConfig},
+};
+use log::{info, warn};
+use trouble_host::prelude::*;
+
+esp_bootloader_esp_idf::esp_app_desc!();
+
+/// Iteration counter, retained across software resets via persistent RTC RAM.
+/// Zeroed once on the initial (power-on) boot, then preserved thereafter.
+#[ram(unstable(rtc_fast, persistent))]
+static mut ITERATION: u32 = 0;
+
+const AP_SSID: &str = "esp-handoff";
+
+/// Max number of connections.
+const CONNECTIONS_MAX: usize = 1;
+/// Max number of L2CAP channels.
+const L2CAP_CHANNELS_MAX: usize = 2; // Signal + att
+
+// GATT Server definition
+#[gatt_server]
+struct Server {
+    battery_service: BatteryService,
+}
+
+/// Battery service
+#[gatt_service(uuid = service::BATTERY)]
+struct BatteryService {
+    /// Battery Level
+    #[descriptor(uuid = descriptors::VALID_RANGE, read, value = [0, 100])]
+    #[descriptor(uuid = descriptors::MEASUREMENT_DESCRIPTION, name = "hello", read, value = "Battery Level", type = &'static str)]
+    #[characteristic(uuid = characteristic::BATTERY_LEVEL, read, notify, value = 10)]
+    level: u8,
+    #[characteristic(uuid = "408813df-5dd4-1f87-ec11-cdb001100000", write, read, notify)]
+    status: bool,
+}
+
+fn init_heap() {
+    #[cfg(feature = "esp32")]
+    {
+        esp_alloc::heap_allocator!(#[ram(reclaimed)] size: 96 * 1024);
+        esp_alloc::heap_allocator!(size: 24 * 1024);
+    }
+    #[cfg(not(feature = "esp32"))]
+    {
+        esp_alloc::heap_allocator!(#[ram(reclaimed)] size: 64 * 1024);
+        esp_alloc::heap_allocator!(size: 64 * 1024);
+    }
+}
+
+#[esp_hal::main]
+async fn main(_s: Spawner) -> ! {
+    esp_println::logger::init_logger_from_env();
+
+    let config = esp_hal::Config::default().with_cpu_clock(CpuClock::max());
+    let peripherals = esp_hal::init(config);
+
+    init_heap();
+
+    let timg0 = TimerGroup::new(peripherals.TIMG0);
+    let sw_int = SoftwareInterruptControl::new(peripherals.SW_INTERRUPT);
+    esp_rtos::start(timg0.timer0, sw_int.software_interrupt0);
+
+    // Read the iteration left behind by the previous boot, then bump it so the
+    // next boot (after our software reset) runs the other radio.
+    let iteration = unsafe { ITERATION };
+    unsafe { ITERATION = iteration.wrapping_add(1) };
+
+    let wifi_phase = iteration % 2 == 0;
+
+    // Pseudo-random active window: 3000..=8000 ms (long enough for a central to
+    // find us and connect during the BLE phase).
+    let rng = Rng::new();
+    let active_ms = 3000 + (rng.random() % 5000);
+
+    info!(
+        "iteration {iteration}: {} phase, active for {active_ms} ms",
+        if wifi_phase { "WIFI" } else { "BLE" }
+    );
+
+    if wifi_phase {
+        run_wifi(peripherals.WIFI, active_ms).await;
+    } else {
+        let connector = BleConnector::new(peripherals.BT, Default::default()).unwrap();
+        let controller: ExternalController<_, 1> = ExternalController::new(connector);
+        run_ble_peripheral(controller, active_ms).await;
+    }
+
+    info!("iteration {iteration} done, software reset");
+    // Give the logger a moment to flush before we reset.
+    Timer::after(Duration::from_millis(50)).await;
+
+    software_reset();
+}
+
+/// Bring up a Wi-Fi SoftAP and keep it beaconing for `active_ms`.
+async fn run_wifi(wifi: esp_hal::peripherals::WIFI<'static>, active_ms: u32) {
+    let ap_config = Config::AccessPoint(AccessPointConfig::default().with_ssid(AP_SSID));
+
+    // Creating the controller in AP mode calls `esp_wifi_start()`, so the radio
+    // starts beaconing immediately.
+    let controller = WifiController::new(
+        wifi,
+        ControllerConfig::default().with_initial_config(ap_config),
+    )
+    .unwrap();
+
+    info!("[WIFI] SoftAP '{AP_SSID}' started, beaconing for {active_ms} ms");
+    Timer::after(Duration::from_millis(active_ms as u64)).await;
+
+    // Tear the radio down before the reset so we exercise the deinit path too.
+    core::mem::drop(controller);
+    info!("[WIFI] controller dropped");
+}
+
+/// Run the BLE battery-service peripheral for `active_ms`, then return so the
+/// caller can software-reset. Mirrors the `bas_peripheral` example, but bounded
+/// by the active window instead of looping forever.
+async fn run_ble_peripheral<C>(controller: C, active_ms: u32)
+where
+    C: Controller,
+{
+    // Using a fixed "random" address can be useful for testing. In real scenarios, one would
+    // use e.g. the MAC 6 byte array as the address (how to get that varies by the platform).
+    let address: Address = Address::random([0xff, 0x8f, 0x1a, 0x05, 0xe4, 0xff]);
+    info!("[BLE] our address = {:?}", address);
+
+    let mut resources: HostResources<DefaultPacketPool, CONNECTIONS_MAX, L2CAP_CHANNELS_MAX> =
+        HostResources::new();
+    let stack = trouble_host::new(controller, &mut resources).set_random_address(address);
+    let Host {
+        mut peripheral,
+        runner,
+        ..
+    } = stack.build();
+
+    info!("[BLE] starting advertising and GATT service");
+    let server = Server::new_with_config(GapConfig::Peripheral(PeripheralConfig {
+        name: "TrouBLE",
+        appearance: &appearance::power_device::GENERIC_POWER_DEVICE,
+    }))
+    .unwrap();
+
+    let app = async {
+        loop {
+            match advertise("Trouble Example", &mut peripheral, &server).await {
+                Ok(conn) => {
+                    // set up tasks when the connection is established to a central, so they don't
+                    // run when no one is connected.
+                    let a = gatt_events_task(&server, &conn);
+                    let b = custom_task(&server, &conn, &stack);
+                    // run until any task ends (usually because the connection has been closed),
+                    // then return to advertising state.
+                    select(a, b).await;
+                }
+                Err(e) => {
+                    warn!("[BLE][adv] error: {:?}", e);
+                    Timer::after(Duration::from_millis(100)).await;
+                }
+            }
+        }
+    };
+
+    // Bound the whole BLE phase by the active window so we always get back to
+    // the software reset and alternate with Wi-Fi.
+    let window = Timer::after(Duration::from_millis(active_ms as u64));
+    select(window, join(ble_task(runner), app)).await;
+    info!("[BLE] active window elapsed");
+}
+
+/// This is a background task that is required to run forever alongside any other BLE tasks.
+async fn ble_task<C: Controller, P: PacketPool>(mut runner: Runner<'_, C, P>) {
+    loop {
+        if let Err(e) = runner.run().await {
+            warn!("[BLE][ble_task] error: {:?}", e);
+            Timer::after(Duration::from_millis(100)).await;
+        }
+    }
+}
+
+/// Stream Events until the connection closes.
+async fn gatt_events_task<P: PacketPool>(
+    server: &Server<'_>,
+    conn: &GattConnection<'_, '_, P>,
+) -> Result<(), Error> {
+    let level = server.battery_service.level;
+    let reason = loop {
+        match conn.next().await {
+            GattConnectionEvent::Disconnected { reason } => break reason,
+            GattConnectionEvent::Gatt { event } => {
+                match &event {
+                    GattEvent::Read(event) => {
+                        if event.handle() == level.handle {
+                            let value = server.get(&level);
+                            info!("[BLE][gatt] read level characteristic: {:?}", value);
+                        }
+                    }
+                    GattEvent::Write(event) => {
+                        if event.handle() == level.handle {
+                            info!("[BLE][gatt] write level characteristic: {:?}", event.data());
+                        }
+                    }
+                    _ => {}
+                };
+                // This step is also performed at drop(), but writing it explicitly is necessary
+                // in order to ensure reply is sent.
+                match event.accept() {
+                    Ok(reply) => reply.send().await,
+                    Err(e) => warn!("[BLE][gatt] error sending response: {:?}", e),
+                };
+            }
+            _ => {} // ignore other Gatt Connection Events
+        }
+    };
+    info!("[BLE][gatt] disconnected: {:?}", reason);
+    Ok(())
+}
+
+/// Create an advertiser to use to connect to a BLE Central, and wait for it to connect.
+async fn advertise<'values, 'server, C: Controller>(
+    name: &'values str,
+    peripheral: &mut Peripheral<'values, C, DefaultPacketPool>,
+    server: &'server Server<'values>,
+) -> Result<GattConnection<'values, 'server, DefaultPacketPool>, BleHostError<C::Error>> {
+    let mut advertiser_data = [0; 31];
+    let len = AdStructure::encode_slice(
+        &[
+            AdStructure::Flags(LE_GENERAL_DISCOVERABLE | BR_EDR_NOT_SUPPORTED),
+            AdStructure::ServiceUuids16(&[[0x0f, 0x18]]),
+            AdStructure::CompleteLocalName(name.as_bytes()),
+        ],
+        &mut advertiser_data[..],
+    )?;
+    let advertiser = peripheral
+        .advertise(
+            &Default::default(),
+            Advertisement::ConnectableScannableUndirected {
+                adv_data: &advertiser_data[..len],
+                scan_data: &[],
+            },
+        )
+        .await?;
+    info!("[BLE][adv] advertising");
+    let conn = advertiser.accept().await?.with_attribute_server(server)?;
+    info!("[BLE][adv] connection established");
+    Ok(conn)
+}
+
+/// Notify the connected central of a counter value and read RSSI every 2 seconds.
+async fn custom_task<C: Controller, P: PacketPool>(
+    server: &Server<'_>,
+    conn: &GattConnection<'_, '_, P>,
+    stack: &Stack<'_, C, P>,
+) {
+    let mut tick: u8 = 0;
+    let level = server.battery_service.level;
+    loop {
+        tick = tick.wrapping_add(1);
+        info!("[BLE][custom_task] notifying connection of tick {}", tick);
+        if level.notify(conn, &tick).await.is_err() {
+            info!("[BLE][custom_task] error notifying connection");
+            break;
+        };
+        if let Ok(rssi) = conn.raw().rssi(stack).await {
+            info!("[BLE][custom_task] RSSI: {:?}", rssi);
+        } else {
+            info!("[BLE][custom_task] error getting RSSI");
+            break;
+        };
+        Timer::after_secs(2).await;
+    }
+}

--- a/qa-test/src/bin/wifi_ble_modem_handoff.rs
+++ b/qa-test/src/bin/wifi_ble_modem_handoff.rs
@@ -1,8 +1,11 @@
 //! Alternating Wi-Fi / BLE radio stress test
 //!
 //! https://github.com/esp-rs/esp-hal/pull/5670
+//!
+//! The easiest way to reproduce the crash is to press Ctrl+R during the Wi-Fi phase.
+//! In the next iteration, when it enters the BLE phase, it should crash immediately.
 //% FEATURES: esp-radio esp-radio/wifi esp-radio/ble esp-radio/unstable esp-hal/unstable
-//% CHIPS: esp32 esp32c3 esp32c5 esp32c6 esp32s3
+//% CHIP_FILTER: wifi_driver_supported && bt_driver_supported && !esp32c2 && !esp32c61
 
 #![no_std]
 #![no_main]
@@ -30,10 +33,38 @@ use trouble_host::prelude::*;
 
 esp_bootloader_esp_idf::esp_app_desc!();
 
-/// Iteration counter, retained across software resets via persistent RTC RAM.
-/// Zeroed once on the initial (power-on) boot, then preserved thereafter.
+/// Iteration counter, retained across software resets.
+///
+/// Most chips keep it in persistent RTC fast RAM: zeroed once on the initial
+/// (power-on) boot, then preserved thereafter. The ESP32-C2 has no RTC fast RAM,
+/// so it stashes the counter in an otherwise-unused RTC_CNTL scratch register
+/// (`STORE6`), which likewise survives a software reset and is cleared only on
+/// power-on reset.
+#[cfg(not(feature = "esp32c2"))]
 #[ram(unstable(rtc_fast, persistent))]
 static mut ITERATION: u32 = 0;
+
+#[cfg(not(feature = "esp32c2"))]
+fn load_iteration() -> u32 {
+    unsafe { ITERATION }
+}
+
+#[cfg(not(feature = "esp32c2"))]
+fn store_iteration(value: u32) {
+    unsafe { ITERATION = value };
+}
+
+#[cfg(feature = "esp32c2")]
+fn load_iteration() -> u32 {
+    esp_hal::peripherals::LPWR::regs().store6().read().bits()
+}
+
+#[cfg(feature = "esp32c2")]
+fn store_iteration(value: u32) {
+    esp_hal::peripherals::LPWR::regs()
+        .store6()
+        .write(|w| unsafe { w.bits(value) });
+}
 
 const AP_SSID: &str = "esp-handoff";
 
@@ -88,8 +119,8 @@ async fn main(_s: Spawner) -> ! {
 
     // Read the iteration left behind by the previous boot, then bump it so the
     // next boot (after our software reset) runs the other radio.
-    let iteration = unsafe { ITERATION };
-    unsafe { ITERATION = iteration.wrapping_add(1) };
+    let iteration = load_iteration();
+    store_iteration(iteration.wrapping_add(1));
 
     let wifi_phase = iteration % 2 == 0;
 

--- a/qa-test/src/bin/wifi_ble_modem_handoff.rs
+++ b/qa-test/src/bin/wifi_ble_modem_handoff.rs
@@ -5,7 +5,7 @@
 //! The easiest way to reproduce the crash is to press Ctrl+R during the Wi-Fi phase.
 //! In the next iteration, when it enters the BLE phase, it should crash immediately.
 //% FEATURES: esp-radio esp-radio/wifi esp-radio/ble esp-radio/unstable esp-hal/unstable
-//% CHIP_FILTER: wifi_driver_supported && bt_driver_supported && !esp32c2 && !esp32c61
+//% CHIP_FILTER: wifi_driver_supported && bt_driver_supported
 
 #![no_std]
 #![no_main]
@@ -36,34 +36,40 @@ esp_bootloader_esp_idf::esp_app_desc!();
 /// Iteration counter, retained across software resets.
 ///
 /// Most chips keep it in persistent RTC fast RAM: zeroed once on the initial
-/// (power-on) boot, then preserved thereafter. The ESP32-C2 has no RTC fast RAM,
-/// so it stashes the counter in an otherwise-unused RTC_CNTL scratch register
-/// (`STORE6`), which likewise survives a software reset and is cleared only on
+/// (power-on) boot, then preserved thereafter. The ESP32-C2 and ESP32-C61 have
+/// no persistent RTC fast RAM, so they stash the counter in an otherwise-unused
+/// always-on scratch register (`STORE6`) - on C2 in `RTC_CNTL`, on C61 in
+/// `LP_AON` - which likewise survives a software reset and is cleared only on
 /// power-on reset.
-#[cfg(not(feature = "esp32c2"))]
+#[cfg(not(any(feature = "esp32c2", feature = "esp32c61")))]
 #[ram(unstable(rtc_fast, persistent))]
 static mut ITERATION: u32 = 0;
 
-#[cfg(not(feature = "esp32c2"))]
 fn load_iteration() -> u32 {
-    unsafe { ITERATION }
+    cfg_select! {
+        feature = "esp32c2" => {
+            esp_hal::peripherals::LPWR::regs().store6().read().bits()
+        },
+        feature = "esp32c61" => {
+            esp_hal::peripherals::LP_AON::regs().store6().read().bits()
+        },
+        _ => unsafe { ITERATION },
+    }
 }
-
-#[cfg(not(feature = "esp32c2"))]
 fn store_iteration(value: u32) {
-    unsafe { ITERATION = value };
-}
-
-#[cfg(feature = "esp32c2")]
-fn load_iteration() -> u32 {
-    esp_hal::peripherals::LPWR::regs().store6().read().bits()
-}
-
-#[cfg(feature = "esp32c2")]
-fn store_iteration(value: u32) {
-    esp_hal::peripherals::LPWR::regs()
-        .store6()
-        .write(|w| unsafe { w.bits(value) });
+    cfg_select! {
+        feature = "esp32c2" => {
+            esp_hal::peripherals::LPWR::regs()
+                .store6()
+                .write(|w| unsafe { w.bits(value) });
+        },
+        feature = "esp32c61" => {
+            esp_hal::peripherals::LP_AON::regs()
+                .store6()
+                .write(|w| unsafe { w.bits(value) });
+        },
+        _ => unsafe { ITERATION = value },
+    }
 }
 
 const AP_SSID: &str = "esp-handoff";
@@ -92,15 +98,15 @@ struct BatteryService {
 }
 
 fn init_heap() {
-    #[cfg(feature = "esp32")]
-    {
-        esp_alloc::heap_allocator!(#[ram(reclaimed)] size: 96 * 1024);
-        esp_alloc::heap_allocator!(size: 24 * 1024);
-    }
-    #[cfg(not(feature = "esp32"))]
-    {
-        esp_alloc::heap_allocator!(#[ram(reclaimed)] size: 64 * 1024);
-        esp_alloc::heap_allocator!(size: 64 * 1024);
+    cfg_select! {
+        feature = "esp32" => {
+            esp_alloc::heap_allocator!(#[ram(reclaimed)] size: 96 * 1024);
+            esp_alloc::heap_allocator!(size: 24 * 1024);
+        },
+        _ => {
+            esp_alloc::heap_allocator!(#[ram(reclaimed)] size: 64 * 1024);
+            esp_alloc::heap_allocator!(size: 64 * 1024);
+        },
     }
 }
 

